lib/runtime-core/src/module.rs

```diff
@@ -164,6 +166,152 @@ impl Module {
     pub fn info(&self) -> &ModuleInfo {
         &self.inner.info
     }
+
+    /// Iterate over the exports that this module provides.
+    ///
+    /// ```
+    /// # use wasmer_runtime_core::module::*;
+    /// # fn example(module: &Module) {
+    /// // We can filter by `ExportKind` to get only certain types of exports.
+    /// // For example, here we get all the names of the functions exported by this module.
+    /// let function_names =
+    ///     module.exports()
+    ///           .filter(|ed| ed.kind == ExportKind::Function)
+    ///           .map(|ed| ed.name)
+    ///           .collect::<Vec<String>>();
+    ///
+    /// // And here we count the number of global variables exported by this module.
+    /// let num_globals =
+    ///     module.exports()
+    ///           .filter(|ed| ed.kind == ExportKind::Global)
+    ///           .count();
+    /// # }
+    /// ```
+    pub fn exports(&self) -> impl Iterator<Item = ExportDescriptor> + '_ {
```

Comment:
I think this should return just the array, not an iterator?

Thoughts?

Reply:
It's definitely a trade off worth considering.

As I see it, returning an array has:
- Pros
  - simpler in the simplest use case
- Cons
  - takes longer at runtime: if we didn't want all the exports or we wanted a different structure, we now have to pay the cost twice
  - forces extra memory use through heap allocation instead of letting the caller decide
  - is less flexible: it forces many uses of this to now be more verbose while allowing the `Vec` case to be handled with a `collect::Vec<_>()`.  It makes the API less composable and interoperable with generic Rust abstractions.

If the API is for users new to Rust then returning a `Vec` makes sense to me, otherwise I think it makes sense to use iterators pretty much everywhere.  The reason being that an iterator is strictly more expressive than a `Vec` (just call `collect` and you get a vec) at the cost of being a concept that people may not be familiar with.

Reply:
Follow up: I spent some time trying to make `imports` return an iterator too and I think it's not possible on stable Rust without a decent amount of work, so that's another point in favor of using `Vec`, it's potentially more stable to large internal breakage because it's less flexible.